README.rst
========
farquaad
========


Named after our incompetent governor, Greg Abbott, farquaad is designed to 
find a vaccine appointment for you. A fitting name to the worst governor
in the country.


Description
===========

There is no system in Texas to place you in a queue, this monitors known 
providers for availability and automatically schedules the first appointment 
within your specified distance.

Currently integrated providers:

- HEB

How this works is you provide details about yourself, a home location and a 
distance you're willig to travel to get the shot(s). farquaad will book your 
appointment automatically. 

Be prepared to travel to the locations within the radius you provide. The 
distance is calculated from the (center) of the home location to the 
destination via geodesic distance measure, not travel time. You can 
ball-park how far you might be willing to travel by checking Google Maps. 
For example, if you're willing to drive from downtown Austin, TX to 
San Antonio, TX, search for HEB stores in San Antonio, TX, and calculate the 
drive time and distance from (for example) 78701. So be warned, travel time 
may be longer or shorter depending on directness of route and traffic.

Note this is subject to change.

Note that some providers may be using Google's reCAPTCHA v3. For those that do,
you will need to have approximately 9 days worth of general browsing with Firefox
to bypass that system. farquaad currently only supports Firefox Selenium driver.

Search Pattern
==============
Several parameters are contribute to the search pattern restriction. That is,
potential vaccination sites are ignored if they fail to meet the search pattern
restrictions specified on the command line. Search pattern restrictions work in
combination so all of them could apply to your search if specified. For example:

-c "Austin,Round Rock" -z 78701,78681

This will restrict it's search to available slots within 50 miles of Austin or 
Round Rock and only within the zipcodes 78701 and 78681. If an appointment in 
78664 is found, it will be ignored (not present in the zipcodes restriction). 
Similarly, if an appointment in Cedar Park is found, it will also be ignored 
(not present in the cities restriction).

-d 50 -H TX,78701

This will restrict it's search to within 50 miles of home (78701). If an appointment
in San Antonio, TX is found, it will be ignored (not within the distance restriction).

Args
====

-H home
    Home location that can be identified by OpenMapQuest

-d miles
    Restriction - Distance in miles you are willing to travel.

-P file
    Patient data file. See data/form.schema.json for JSON schema validation.
    Note that this file is not currently validated against the schema.

-c city,city,...
    Restriction - List of cities to restrict your search.

-z zip,zip,...
    Restriction - List of zipcodes to restrict your search.

-v
    Info level verbosity logged

-vv
    Debug level verbosity logged

--version
    Print the current version


CHANGELOG.rst
=========
Changelog
=========

Version 0.1
===========

- initial revision for local install


tests/conftest.py
"""
    Dummy conftest.py for farquaad.

    If you don't know what this is for, just leave it empty.
    Read more about conftest.py under:
    - https://docs.pytest.org/en/stable/fixture.html
    - https://docs.pytest.org/en/stable/writing_plugins.html
"""

# import pytest


tests/test_cli.py
import pytest

from farquaad.cli import fib, main

__author__ = "Jason Switzer"
__copyright__ = "Jason Switzer"
__license__ = "MIT"


def test_fib():
    """API Tests"""
    assert fib(1) == 1
    assert fib(2) == 1
    assert fib(7) == 13
    with pytest.raises(AssertionError):
        fib(-10)


def test_main(capsys):
    """CLI Tests"""
    # capsys is a pytest fixture that allows asserts agains stdout/stderr
    # https://docs.pytest.org/en/stable/capture.html
    main(["7"])
    captured = capsys.readouterr()
    assert "The 7-th Fibonacci number is 13" in captured.out


src/farquaad/distributors/heb.py
from urllib.request import urlopen
import json

from . import Distributor
from ..base import Page

class HEB(Distributor):

    def __init__(self, driver, logger):
        super().__init__("H.E.B.", driver, logger)
        self.statusUrl = "https://heb-ecom-covid-vaccine.hebdigital-prd.com/vaccine_locations.json"
        self.page_appointment = AppointmentForm(driver, logger)
        self.page_patient = PatientForm(driver, logger)
        self.page_confirm = Page(driver, logger)
    
    def process(self, openSlot, patientData):
        self._logger.info("HEB processor; processing Appointment, Patient, Confirmation pages")
        
        shouldContinue = True
        if not self.page_appointment.load(openSlot["url"]):
            return False
        if not self.page_appointment.populate(patientData):
            return False
        self.page_appointment.capture("/tmp/farquaad/heb_appointment_page.png")
        self.page_appointment.proceed()

        if not self.page_patient.load():
            return False
        if not self.page_patient.populate(patientData):
            return False
        self.page_patient.capture("screenshots/heb_appointment_page.png")
        self.page_patient.proceed()

        # this is just a plain page, nothing to do here
        self.page_confirm.capture("screenshots/heb_appointment_page.png")
        return True
    
    def checkAvailability(self, restrictions):
        self._logger.info("Checking HEB API for availability")
        locations = json.loads(urlopen(self.statusUrl).read())['locations']

        filtered = [x for x in locations if x["openTimeslots"] > 0]
        return filtered

class AppointmentForm(Page):

    def __init__(self, driver, logger):
        super().__init__(driver, logger)
        self.widgets = {
            "combo-manufacturer": "//lightning-combobox[1]",
            "item-manufacturer": {
                "Any": "//lightning-combobox[1]//*[@data-value='Any']",
                "J&J/Janssen": "//lightning-combobox[1]//*[@data-value='Janssen']",
                "Moderna": "//lightning-combobox[1]//*[@data-value='Moderna']",
                "Pfizer": "//lightning-combobox[1]//*[@data-value='Pfizer']",
            },
            "combo-date": "//lightning-combobox[2]",
            "items-date": "//lightning-combobox[2]//*//lightning-base-combobox-item",
            "item-date": [
                "//lightning-combobox[2]//*//lightning-base-combobox-item[10]/span[2]/span"
            ],
            "combo-time": "//lightning-combobox[3]",
            "items-time": "//lightning-combobox[3]//*//lightning-base-combobox-item",
            "item-time": [
                "//lightning-combobox[3]//*//lightning-base-combobox-item[1]/span[2]/span"
            ],
            "button-continue": "//lightning-button/button",
            "sucka": "//*[contains(text(), 'There are no available time slots.')]"
        }

    def populate(self, patientData):
        self._logger.info("Populating the Appointment form")
        #self._logger.info("Form fields: " + json.dumps(self.widgets, indent=4))

        # for each manufacturer the patient will accept, find the first available date and time
        # problem here is this form is dynamic, uses shadow DOM, and errors out _lots_ of ways
        for m in patientData["manufacturer"]:
            self.scrollTo(self.widgets["combo-manufacturer"], click=True)
            self._logger.info(f"Selecting shot: {m}")
            self.clickOn(self.widgets["item-manufacturer"][m])
            try:
                self.wait(self.widgets["combo-date"], 1)
                self._logger.info(f"Manufacturer successfully selected as {m}")
            except:
                self._logger.info("Appointment date combobox is unavailable, continuing")
                continue
            self.scrollTo(self.widgets["combo-date"], click=True)
            dateOptions = self.findAll(self.widgets["items-date"])
            if len(dateOptions) <= 0:
                self._logger.info(f"No dates found for {m}, continuing")
                continue
            self._logger.info(f"Found {len(dateOptions)} date options for {m}")
            for d in dateOptions:
                d.click()
                try:
                    self.wait(self.widgets["combo-time"], 1)
                    self._logger.info(f"Date successfully selected for {m}")
                except:
                    self._logger.info("Appointment time combobox is unavailable, continuing")
                    continue
                self.scrollTo(self.widgets["combo-time"], click=True)
                timeOptions = self.findAll(self.widgets["items-time"])
                if len(timeOptions) <= 0:
                    self._logger.info("No time slots found, continuing")
                    continue
                self._logger.info(f"Found {len(timeOptions)} time slot options for {m}")
                for t in timeOptions:
                    t.click()
                    try:
                        self.wait(self.widgets["button-continue"], 1)
                        self._logger.info(f"Time slot successfully selected for {m}")
                    except:
                        self._logger.info("Continue button is unavailable, continuing")
                        continue
                    if self.isPresent(self.widgets["button-continue"]):
                        return True
        return False
    
    def load(self, url):
        self._logger.info("Loading the Appointment Form")
        return super().load(url, self.widgets["combo-manufacturer"]) and not self.isPresent(self.widgets["sucka"])

    def proceed(self):
        self._logger.info("Proceeding from the Appointment Form to the next page")
        self.scrollTo(self.widgets["button-continue"], click=True)

class PatientForm(Page):
    def __init__(self, driver, logger):
        super().__init__(driver, logger)
        self.widgets = {
            "text-firstname": "//input[contains(@name, 'First_Name')]",
            "text-lastname": "//input[contains(@name, 'Last_Name')]",
            "text-email": "//input[contains(@name, 'Email_Address')]",
            "text-birthdate": "//input[@data-type='birthdate']",
            "text-phone": "//input[@data-type='tel']",
            "text-peoplesoft": "//lightning-input[4]//*//input",
            "combo-provider": "//lightning-combobox[1]",
            "item-provider": {
                "yes": "//lightning-combobox[1]//*//lightning-base-combobox-item[1]/span[2]/span",
                "no": "//lightning-combobox[1]//*//lightning-base-combobox-item[2]/span[2]/span"
            },
            "text-provider": "//lightning-input[5]//*//input",
            "text-providerid": "//lightning-input[6]//*//input",
            "text-groupnumber": "//lightning-input[6]//*//input",
            "combo-phase": "//lightning-combobox[2]",
            "item-phase": {
                "phase1": "//lightning-combobox[2]//*//lightning-base-combobox-item[1]/span[2]/span",
                "phase1b": "//lightning-combobox[2]//*//lightning-base-combobox-item[2]/span[2]/span",
                "phase1bplus": "//lightning-combobox[2]//*//lightning-base-combobox-item[3]/span[2]/span"
                "phase1c": "//lightning-combobox[2]//*//lightning-base-combobox-item[4]/span[2]/span"
                #"phase2": "//lightning-combobox[2]//*//lightning-base-combobox-item[5]/span[2]/span"
            },
            "button-schedule": "//lightning-button[2]/button",
        }

    def populate(self, patientData):
        self._logger.info("Populating the Patient form")

        # first the mandatory text fields
        self.scrollTo(self.widgets["text-firstname"])
        self.sendKeys(self.widgets["text-firstname"], patientData["firstname"])
        self.scrollTo(self.widgets["text-lastname"])
        self.sendKeys(self.widgets["text-lastname"], patientData["lastname"])
        self.scrollTo(self.widgets["text-email"])
        self.sendKeys(self.widgets["text-email"], patientData["email"])
        self.scrollTo(self.widgets["text-birthdate"])
        self.sendKeys(self.widgets["text-birthdate"], patientData["birthdate"])
        self.scrollTo(self.widgets["text-phone"])
        self.sendKeys(self.widgets["text-phone"], patientData["phone"])

        self.scrollTo(self.widgets["text-peoplesoft"], click=True)
        if "peoplesoft" in patientData.keys() and patientData["peoplesoft"]:
            self.sendKeys(self.widgets["text-peoplesoft"], self.widgets["peoplesoft"])

        self.scrollTo(self.widgets["combo-provider"], click=True)
        if "provider" in patientData.keys() and patientData["provider"]:
            self.clickOn(self.widgets["item-provider"]["yes"])
            #click to force the combo closed, just in case
            self.scrollTo(self.widgets["text-provider"], click=True)
            self.sendKeys(self.widgets["text-provider"], patientData["provider"])
        else:
            self.clickOn(self.widgets["item-provider"]["no"])

        #clear the state, make sure none of the drop downs are open
        self.scrollTo(self.widgets["text-providerid"], click=True)
        if "providerid" in patientData.keys() and patientData["providerid"]:
            self.sendKeys(self.widgets["text-providerid"], self.widgets["providerid"])
        
        self.scrollTo(self.widgets["text-groupnumber"], click=True)
        if "groupnumber" in patientData.keys() and patientData["groupnumber"]:
            self.sendKeys(self.widgets["text-groupnumber"], self.widgets["groupnumber"])

        self.scrollTo(self.widgets["combo-phase"], click=True)
        self.clickOn(self.widgets["item-phase"][patientData["certify"].lower()])
        return True

    def load(self):
        self._logger.info("Loading the Appointment Form")
        #can't directly load, just confirm the page loaded from the prior proceed
        try:
            self.wait(self.widgets["button-schedule"])
        except:
            return False
        return self.isPresent(self.widgets["button-schedule"])
    
    def proceed(self):
        self._logger.info("Proceeding from the Patient Form to the next page")
        self.scrollTo(self.widgets["button-schedule"], click=True)

src/farquaad/distributors/__init__.py
import sys
from time import sleep
from ..base import Filter

class Distributor:
    def __init__(self, name, driver, logger):
        self._name = name
        self._driver = driver
        self._logger = logger
        self._logger.info(f"Initialized distributor registration journey: {self._name}")

    def checkAvailability(self, restrictions):
        self._logger.info(f"Checking distributor registration availability: {self._name}")

    def monitor(self, home, restrictions, delay):
        self._logger.info(f"Monitoring distributor registration availability: {self._name}")
        availability = None
        refine = Filter(home, restrictions)
        while not availability:
            sleep(delay)
            availability = self.checkAvailability(restrictions)
            availability = refine.apply(availability)
            availability = sorted(availability, key=lambda slot: slot["distance"])
        return availability
    
    def process(self, openSlot, patientData):
        self._logger.info(f"Processing distributor registration journey: {self._name}")

    def schedule(self, availability, patientdata):
        self._logger.info(f"Searching for and scheduling distributor registration journey: {self._name}")
        booked = False
        for appt in availability:
            booked = self.process(appt, patientdata)
            if booked:
                return appt

    def finalize(self):
        self._driver.quit()
        self._logger.info(f"Finalizing registration distributor: {self._name}")



src/farquaad/cli.py
import argparse
import json
import logging
import sys

from selenium import webdriver
from selenium.webdriver.firefox.options import Options

from .distributors.heb import HEB
from .base import Restrictions

from farquaad import __version__

__author__ = "Jason Switzer"
__copyright__ = "Jason Switzer"
__license__ = "MIT"

_logger = logging.getLogger(__name__)

# ---- Python API ----
# The functions defined in this section can be imported by users in their
# Python scripts/interactive interpreter, e.g. via
# `from farquaad.cli import heb`,
# when using this Python module as a library.

def heb(driver=None, logger=None):
    return HEB(driver, logger)

# ---- CLI ----
# The functions defined in this section are wrappers around the main Python
# API allowing them to be called directly from the terminal as a CLI
# executable/script.

def parse_args(args):
    parser = argparse.ArgumentParser(description="Automated Vaccine scheduler")
    parser.add_argument(
        "--version",
        action="version",
        version="farquaad {ver}".format(ver=__version__),
    )
    parser.add_argument(
        "-v",
        "--verbose",
        dest="loglevel",
        help="set loglevel to INFO",
        action="store_const",
        const=logging.INFO,
    )
    parser.add_argument(
        "-vv",
        "--very-verbose",
        dest="loglevel",
        help="set loglevel to DEBUG",
        action="store_const",
        const=logging.DEBUG,
    )
    parser.add_argument(
        "-c",
        "--cities",
        dest="cities",
        help="Cities to restrict appointment availability",
        nargs="+",
    )
    parser.add_argument(
        "-d",
        "--distance",
        dest="distance",
        help="Maximum distance (miles) from home (requires --home)",
        type=float,
    )
    parser.add_argument(
        "-H",
        "--home",
        dest="home",
        help="Home location, typically includes zipcode, address, lat/lon, city, etc - uses OpenStreetMap (requires --distance)",
    )
    parser.add_argument(
        "-z",
        "--zip-codes",
        dest="zipcodes",
        help="Zipcodes to restrict appointment availability",
        nargs="+"
    )
    parser.add_argument(
        "-t",
        "--time-delay",
        dest="timedelay",
        help="Time delay between availability checks in seconds",
        type=float,
        default=10.0,
    )
    parser.add_argument(
        "-P",
        "--patient-data",
        dest="patientdata",
        help="Patient automation data file - see form.json for a sample",
        type=argparse.FileType('r'), 
    )
    return parser.parse_args(args)

def setup_logging(loglevel):
    """Setup basic logging

    Args:
      loglevel (int): minimum loglevel for emitting messages
    """
    logformat = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    logging.basicConfig(
        level=loglevel, stream=sys.stdout, format=logformat, datefmt="%Y-%m-%d %H:%M:%S"
    )

def main(args):
    args = parse_args(args)
    setup_logging(args.loglevel)
    patientData = json.load(args.patientdata)
    restricted = Restrictions(cities=args.cities, zipcodes=args.zipcodes, distance=args.distance)
    _logger.debug("Initializing the default web driver")
    
    # let's create the driver here so we can reuse it
    options = Options()
    options.headless = False
    driver = webdriver.Firefox(options=options)

    # find a store, grab a slot, cleanup
    henryebutts = heb(driver=driver, logger=_logger)
    appt = None
    while not appt:
        possibleSites = henryebutts.monitor(args.home, restricted, args.timedelay)
        print(possibleSites)
        appt = henryebutts.schedule(possibleSites, patientData)
        if appt:
            _logger.info(f"Appointment scheduled at {appt['name']}")
            break
        #if we didn't break out, all of the slots found booked too quick, start over
    henryebutts.finalize()
    _logger.info("Auto-registration complete, heck the provided email for the appointment")

def run():
    """Calls :func:`main` passing the CLI arguments extracted from :obj:`sys.argv`

    This function can be used as entry point to create console scripts with setuptools.
    """
    main(sys.argv[1:])

if __name__ == "__main__":
    # ^  This is a guard statement that will prevent the following code from
    #    being executed in the case someone imports this file instead of
    #    executing it as a script.
    #    https://docs.python.org/3/library/__main__.html

    # After installing your project with pip, users can also run your Python
    # modules as scripts via the ``-m`` flag, as defined in PEP 338::
    #
    #     python -m farquaad.cli -H TX,78701 -d 50 -P data/patient.json
    #
    run()
